### PR TITLE
add optional http basic auth to pinot broker

### PR DESCRIPTION
**Pull request by apucher**
## Description
Added support for an optional HTTP basic auth provider to pinot-broker, which enables user- and table-level authentication of incoming queries.

HTTP basic auth configuration uses pinot's regular properties format. For example:
> pinot.broker.access.control.class=org.apache.pinot.broker.broker.BasicAuthAccessControlFactory
> pinot.broker.access.control.principals=admin,user
> pinot.broker.access.control.principals.admin.password=verysecret
> pinot.broker.access.control.principals.user.password=secret
> pinot.broker.access.control.principals.user.tables=myUserTable,myOtherUserTable

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [x] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

## Release Notes

Added support for an optional HTTP basic auth provider to pinot-broker, which enables user- and table-level authentication of incoming queries.

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  /**
+   * Access Control using header-based basic http authentication
+   */
+  private static class BasicAuthAccessControl implements AccessControl {
+    private final Map<String, BasicAuthPrincipal> _principals;
+
+    public BasicAuthAccessControl(Collection<BasicAuthPrincipal> principals) {
+      this._principals = principals.stream().collect(Collectors.toMap(BasicAuthPrincipal::getToken, p -> p));
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:92]: (nit) remove `this.` Same for other places

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Container object for basic auth principal
+   */
+  private static class BasicAuthPrincipal {
+    private final String _name;
+    private final String _token;
+    private final Set<String> _tables;
+
+    public BasicAuthPrincipal(String name, String token, Set<String> tables) {
+      this._name = name;
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:130]: (nit) Remove `this.`

File: pinot-broker/src/test/java/org/apache/pinot/broker/broker/BasicAuthAccessControlTest.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class BasicAuthAccessControlTest {
+    private static final String TOKEN_USER = "Basic dXNlcjpzZWNyZXQ"; // user:secret
```
Comment (Jackie-Jiang) [inline pinot-broker/src/test/java/org/apache/pinot/broker/broker/BasicAuthAccessControlTest.java:36]: Fix the indentation and the import

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java [hunk 4]
```diff
@@ -210,7 +214,14 @@ public String getQueryResponse(String query, String traceEnabled, String queryOp
     String url = getQueryURL(protocol, hostNameWithPrefix.substring(hostNameWithPrefix.indexOf("_") + 1),
         String.valueOf(port), querySyntax);
     ObjectNode requestJson = getRequestJson(query, traceEnabled, queryOptions, querySyntax);
-    return sendRequestRaw(url, query, requestJson);
+
+    // forward client-supplied headers
+    Map<String, String> headers = httpHeaders.getRequestHeaders().entrySet().stream()
+        .filter(entry -> !entry.getValue().isEmpty())
+        .map(entry -> Pair.of(entry.getKey(), entry.getValue().get(0)))
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:221]: Is it okay to only keep the first value?

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    }
+  }
+
+  private static String toToken(String name, String password) {
+    String identifier = String.format("%s:%s", name, password);
+    return normalizeToken(String.format("Basic %s", Base64.getEncoder().encodeToString(identifier.getBytes())));
+  }
+
+  private static String normalizeToken(String token) {
+    if (token == null) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:154]: token should never be `null`?

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      return null;
+    }
+    return token.trim().replace("=", "");
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:157]: Why do we need to remove the `=`?
Besides, `StringUtils.remove(token.trim(), '=')` is faster.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    public String getToken() {
+      return _token;
+    }
+  }
+
+  private static String toToken(String name, String password) {
+    String identifier = String.format("%s:%s", name, password);
+    return normalizeToken(String.format("Basic %s", Base64.getEncoder().encodeToString(identifier.getBytes())));
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:150]: Should we use `UTF-8` here?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java (hunk 4 above)
Comment (apucher) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:221]: For our purposes, yes. Pinot's handling of http headers overall isn't too glorious.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 6 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:157]: the implementation of base64 encoding across different libraries varies. "=" is a placeholder to align the number of output bytes, but some libs strip it or align differently, hence, for the same input we may end up with "aaa", "aaa=", "aaa==" and so on. They're ultimately equivalent, but hash-matching may fail if it isn't normalized.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 5 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:154]: technically, null-value is possible in header. see line 102.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 6 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:157]: btw StringUtils literally calls replace :P

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 1 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:92]: done

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 2 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:130]: done

File: pinot-broker/src/test/java/org/apache/pinot/broker/broker/BasicAuthAccessControlTest.java (hunk 3 above)
Comment (apucher) [reply pinot-broker/src/test/java/org/apache/pinot/broker/broker/BasicAuthAccessControlTest.java:36]: done

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 7 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:150]: done

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 5 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:163]: If it returns `null`, will `.map(_principals::get)` on line 103 throw NPE?
If we allow `null` here, suggest annotate token with `@Nullable`

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.pinot.broker.api.AccessControl;
+import org.apache.pinot.broker.api.HttpRequesterIdentity;
+import org.apache.pinot.broker.api.RequesterIdentity;
+import org.apache.pinot.common.request.BrokerRequest;
+import org.apache.pinot.spi.env.PinotConfiguration;
+
+import java.nio.charset.StandardCharsets;
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:29]: Can you follow the instructions here to set up the IDE and auto-reformat the code: https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java [hunk 9]
```diff
@@ -29,18 +29,22 @@
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:32]: Some imports are unused?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 8 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:29]: I did this already. Auto-formatted the code again, just added a few line breaks. What are you looking for, specifically?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java (hunk 9 above)
Comment (apucher) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:32]: removed

**Comment by apucher** [comment]
re the first comment (which I somehow can't respond to inline):
`.map(_principals::get)` is fine if the key, i.e. the token, is null. There's even a unit tests for this.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 8 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:30]: Weird.. The setting should not split the imports into 2 parts

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+   *
+   * @param token raw token
+   * @return normalized token
+   */
+  @Nullable
+  private static String normalizeToken(String token) {
+    if (token == null) {
+      return null;
+    }
+    return StringUtils.remove(token.trim(), "=");
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:168]:
```suggestion
    return StringUtils.remove(token.trim(), '=');
```

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java (hunk 8 above)
Comment (apucher) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/BasicAuthAccessControlFactory.java:29]: meh. fixed

**Pull request merged**